File: sdk/experimental/rs/src/telemetry/mode.rs
//! Mode enum + env precedence.
//!
//! Mirrors `go/runtime/telemetry/mode.go` semantics: three tiers (off,
//! anon, full), <APP>_TELEMETRY_MODE wins over KIT_TELEMETRY_MODE.
//! Unknown tokens map to `Mode::Off + false` so callers can distinguish
//! "operator typoed" from "operator opted out". The SDK does NOT carry
//! the Go atomic-global / one-shot env-read machinery — each
//! `resolve_mode` call reads the environment fresh; Rust adopters who
//! need stickiness should cache the result themselves.

use serde::{Deserialize, Serialize};
use std::env;

/// Telemetry emission tier.
///
/// - `Off`: default; emit is a no-op.
/// - `Anon`: anonymous payload (installation_id + command_path +
///   exit_code + duration_ms + occurred_at + kit_version + sdk).
/// - `Full`: `Anon` plus args + flags, both AFTER redaction.
#[derive(Copy, Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum Mode {
    Off,
    Anon,
    Full,
}

/// Parse a Mode token (case-insensitive, leading/trailing whitespace
/// tolerated). Empty/missing input parses as `(Off, true)` so an unset
/// env var doesn't masquerade as an error. Unknown tokens return
/// `(Off, false)` so callers can detect typos.
pub fn parse_mode(s: Option<&str>) -> (Mode, bool) {
    match s {
        None => (Mode::Off, true),
        Some(raw) => {
            let lower = raw.trim().to_lowercase();
            match lower.as_str() {
                "" | "off" => (Mode::Off, true),
                "anon" => (Mode::Anon, true),
                "full" => (Mode::Full, true),
                _ => (Mode::Off, false),
            }
        }
    }
}

/// Read the `KIT_APP_PREFIX` env var, trimmed + upper-cased. Empty
/// string when unset or whitespace-only.
fn resolve_app_prefix() -> String {
    env::var("KIT_APP_PREFIX")
        .unwrap_or_default()
        .trim()
        .to_uppercase()
}

/// Resolve the active Mode from environment in precedence order:
///   1. `<APP>_TELEMETRY_MODE` (if `KIT_APP_PREFIX` is set)
///   2. `KIT_TELEMETRY_MODE`
///   3. `Mode::Off` (default)
///
/// Invalid tokens at any level fall through to the next source; if all
/// sources are unset/invalid the result is `Mode::Off`. This mirrors
/// the Go-side `readEnvMode` precedence (mode.go) without the atomic
/// global — Rust adopters typically resolve once at startup.
pub fn resolve_mode() -> Mode {
    let app = resolve_app_prefix();
    if !app.is_empty() {
        if let Ok(v) = env::var(format!("{}_TELEMETRY_MODE", app)) {
            let (m, ok) = parse_mode(Some(&v));
            if ok {
                return m;
            }
        }
    }
    if let Ok(v) = env::var("KIT_TELEMETRY_MODE") {
        let (m, ok) = parse_mode(Some(&v));
        if ok {
            return m;
        }
    }
    Mode::Off
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parse_mode_known_tokens() {
        assert_eq!(parse_mode(Some("off")), (Mode::Off, true));
        assert_eq!(parse_mode(Some("OFF")), (Mode::Off, true));
        assert_eq!(parse_mode(Some("anon")), (Mode::Anon, true));
        assert_eq!(parse_mode(Some("ANON")), (Mode::Anon, true));
        assert_eq!(parse_mode(Some(" full ")), (Mode::Full, true));
        assert_eq!(parse_mode(Some("Full")), (Mode::Full, true));
    }

    #[test]
    fn parse_mode_empty_or_missing() {
        assert_eq!(parse_mode(None), (Mode::Off, true));
        assert_eq!(parse_mode(Some("")), (Mode::Off, true));
        assert_eq!(parse_mode(Some("   ")), (Mode::Off, true));
    }

    #[test]
    fn parse_mode_unknown_returns_off_and_false() {
        assert_eq!(parse_mode(Some("garbage")), (Mode::Off, false));
        assert_eq!(parse_mode(Some("anonymous")), (Mode::Off, false));
        assert_eq!(parse_mode(Some("on")), (Mode::Off, false));
    }

    // resolve_mode is exercised via tests/mode.rs — those tests run in
    // their own process so env-var manipulation can't race the api_test
    // tokio runtime. std::env::set_var is unsafe under threading.
}

File: sdk/py/hop_top_kit/telemetry/mode.py
"""Telemetry mode enum + env-precedence resolver.

Mirrors `go/runtime/telemetry/mode.go`. SDK-side resolution honors:
  1. <APP>_TELEMETRY_MODE (when KIT_APP_PREFIX is set)
  2. KIT_TELEMETRY_MODE
  3. Mode.OFF (default)

DO_NOT_TRACK and KIT_TELEMETRY_CONSENT are NOT consulted here — that's the
consent layer's job (see consent.py). This module is pure mode resolution.
The client combines mode + consent before any emission.
"""

from __future__ import annotations

import os
from enum import StrEnum
from typing import Optional


class Mode(StrEnum):
    """Telemetry emission mode. String-valued for ergonomic comparison."""

    OFF = "off"
    ANON = "anon"
    FULL = "full"


def parse_mode(s: Optional[str]) -> tuple[Mode, bool]:
    """Parse a mode string.

    Returns (mode, ok). Empty / None maps to (OFF, True) — treated as
    "no opinion, default off". Unknown maps to (OFF, False).
    """
    if s is None or s == "":
        return Mode.OFF, True
    low = s.strip().lower()
    for m in Mode:
        if m.value == low:
            return m, True
    return Mode.OFF, False


def _resolve_app_prefix(env: dict) -> str:
    """SDKs adopt the app prefix from KIT_APP_PREFIX (rarely set; KIT-only mostly)."""
    return env.get("KIT_APP_PREFIX", "").strip().upper()


def resolve_mode(env: Optional[dict] = None) -> Mode:
    """Resolve the current telemetry mode per precedence.

    Precedence (highest wins):
      1. <APP>_TELEMETRY_MODE (if KIT_APP_PREFIX is set)
      2. KIT_TELEMETRY_MODE
      3. Mode.OFF (default)

    Malformed values fall through to the next layer rather than raising;
    fully malformed env → OFF.
    """
    env = env if env is not None else dict(os.environ)

    app = _resolve_app_prefix(env)
    if app:
        v = env.get(f"{app}_TELEMETRY_MODE", "").strip()
        if v:
            m, ok = parse_mode(v)
            if ok:
                return m

    v = env.get("KIT_TELEMETRY_MODE", "").strip()
    if v:
        m, ok = parse_mode(v)
        if ok:
            return m

    return Mode.OFF
